lahja/trio/endpoint.py

```diff
@@ -0,0 +1,905 @@
+
+    async def wait_stopped(self) -> None:
+        await self._stopped.wait()
+
+    @asynccontextmanager
+    async def run(self) -> AsyncGenerator[ConnectionAPI, None]:
+        async with trio.open_nursery() as nursery:
+            await self._start(nursery)
+            try:
+                yield self
+            finally:
+                await self._stop()
+
+    async def _start(self, nursery: trio_typing.Nursery) -> None:
+        if self.is_running:
+            raise LifecycleError("Connection is already running")
+        elif self.is_stopped:
+            raise LifecycleError("Connection has already been run and stopped")
+
+        nursery.start_soon(self._run)
+        # this is a trade-off between allowing flexibility in how long it takes
+        # for a `Runnable` to start and getting good errors in the event that
+        # the `_run` method forgets to set the `_running` event.
+        with trio.fail_after(2):
+            await self.wait_started()
+
+    async def _stop(self) -> None:
+        if self.is_stopped:
+            return
+        self._stopped.set()
+        await self._cleanup()
+
+    async def _run(self) -> None:
+        """
+        This method buffers the underlying socket, reading larger quantities of
+        data from the socket at a time and then buffering the data internally
+        to pull of the individual messages.
+        """
+        self.logger.debug("Starting TrioConnection")
+        buffer = bytearray()
+        self._running.set()
+
+        while not self.is_stopped:
+            while len(buffer) < 4:
+                try:
+                    data = await self._socket.receive_some(2048)
+                except (trio.ClosedResourceError, trio.BrokenResourceError):
+                    await self._stop()
+                    return
+
+                if data == b"":
+                    await self._stop()
+                    return
+
+                buffer.extend(data)
+
+            t_size = 4 + int.from_bytes(buffer[:4], "little")
+
+            while len(buffer) < t_size:
+                data = await self._socket.receive_some(2048)
+                if data == b"":
+                    raise RemoteDisconnected()
+                buffer.extend(data)
+
+            msg = cast(Message, pickle.loads(buffer[4:t_size]))
+            await self._msg_send_channel.send(msg)
+            buffer = buffer[t_size:]
+
+    async def _cleanup(self) -> None:
+        await self._msg_send_channel.aclose()
+        await self._socket.aclose()
+
+    #
+    # Connection API
+    #
+    @classmethod
+    async def connect_to(cls, path: pathlib.Path) -> "TrioConnection":
+        socket = await trio.open_unix_socket(str(path))
+        cls.logger.debug("Opened connection to %s: %s", path, socket)
+        return cls(socket)
+
+    async def send_message(self, message: Msg) -> None:
+        msg_data = pickle.dumps(message)
+        size = len(msg_data)
+        async with self._lock:
```

Comment:
I guess this lock is because `send_all` can switch away when there's backpressure and then throw an exception if we try to use it reentrantly. I was surprised that trio didn't lock for us but then found [this thread](https://github.com/python-trio/trio/issues/328). 

I think future Brian's will have an easier time reading this if the lock was called something like `_write_lock` and if there to be a comment saying what it's doing, maybe even with a link to the above thread.

Reply:
:+1: